### ISSUE TITLE
Crash on clipboard restoration

### ISSUE THREAD

**Issue by LukeLR** [autokey/autokey#226]
## Summary

When using a phrase that replaces an abbreviation using one of the clipboard methods, AutoKey crashes in 95% of all cases after entering the abbreviation twice. The first occurrence gets replaced, but the second one only gets deleted and not replaced.

## Steps to Reproduce (if applicable)

1. Create a new phrase containing something (a single letter is sufficient)
2. Add an abbreviation to that phrase
3. Select one of the methods using the clipboard for replacement.
4. Open a text window (e.g. mousepad) and enter the abbreviation of the phrase twice.

## Expected Results

The phrase works as expected and the abbreviation gets replaced every single time.

## Actual Results

In most cases, AutoKey crashes on the second occurence of the abbreviation. It does not respond to ^C on the command line, but `kill -s 15 <pid of autokill-gtk>` is sufficient. After AutoKey has crashed, no keyboard input is passed to X, unless I switch workspaces once (using i3, scrolling on the menu bar).

When running with `autokey-gtk --verbose`, I get the following logs when reproducing the crash:
```
2018-12-05 07:02:31,958 DEBUG - service - Raw key: 'b', modifiers: [], Key: b
2018-12-05 07:02:31,958 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
2018-12-05 07:02:31,958 DEBUG - service - No phrase/script matched hotkey
2018-12-05 07:02:31,958 DEBUG - service - Input stack at end of handle_keypress: deque(['b'], maxlen=150)
2018-12-05 07:02:32,127 DEBUG - service - Raw key: ' ', modifiers: [], Key:  
2018-12-05 07:02:32,128 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
2018-12-05 07:02:32,128 DEBUG - service - No phrase/script matched hotkey
2018-12-05 07:02:32,132 DEBUG - iomediator - Send via clipboard
2018-12-05 07:02:32,133 DEBUG - service - Input stack at end of handle_keypress: deque([], maxlen=150)
2018-12-05 07:02:32,133 DEBUG - interface - Sending string via clipboard: a 
2018-12-05 07:02:32,133 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2018-12-05 07:02:32,135 DEBUG - interface - Sending via clipboard enqueued.
2018-12-05 07:02:32,135 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-12-05 07:02:32,137 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-12-05 07:02:32,139 DEBUG - iomediator - Send via event interface
2018-12-05 07:02:32,139 DEBUG - interface - Send modified key: modifiers: ['<shift>'] key: <insert>
2018-12-05 07:02:32,242 ERROR - interface - Error in X event loop thread
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 230, in __eventLoop
    method(*args)
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 626, in _restore_clipboard_text
    self.clipboard.text = backup
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 161, in text
    self._clipboard.set_text(new_content, -1)
TypeError: Argument 1 does not allow None as a value
2018-12-05 07:02:33,039 DEBUG - service - Raw key: 'b', modifiers: [], Key: b
2018-12-05 07:02:33,040 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
2018-12-05 07:02:33,040 DEBUG - service - No phrase/script matched hotkey
2018-12-05 07:02:33,041 DEBUG - service - Input stack at end of handle_keypress: deque(['b'], maxlen=150)
2018-12-05 07:02:33,318 DEBUG - service - Raw key: ' ', modifiers: [], Key:  
2018-12-05 07:02:33,319 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
2018-12-05 07:02:33,319 DEBUG - service - No phrase/script matched hotkey
2018-12-05 07:02:33,322 DEBUG - iomediator - Send via clipboard
2018-12-05 07:02:33,323 DEBUG - service - Input stack at end of handle_keypress: deque([], maxlen=150)
2018-12-05 07:02:33,325 DEBUG - interface - Sending string via clipboard: a 
2018-12-05 07:02:33,327 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2018-12-05 07:02:33,328 DEBUG - interface - Sending via clipboard enqueued.
2018-12-05 07:02:33,329 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
2018-12-05 07:02:33,333 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
```

## Version

AutoKey version: 0.95.4

Used GUI (Gtk, Qt, or both): Gtk

If the problem is known to be present in more than one version, please list all of those.

Installed via: pacman

Distro: Arch Linux

## Notes

Using a script like this instead of a phrase works:
```
from subprocess import Popen, PIPE                                                                                                                                        
def paste_character(symbol):                                                                                                                                                   
    c = Popen(['xclip', '-selection', 'clipboard'], stdin=PIPE)                                                                                                           
    c.communicate(symbol.encode('utf-8'))                                                                                                                                 
    keyboard.send_keys('<ctrl>+v')

paste_character("☺")
```

This might be somewhat related to #225, but with the following differences:

- If the crash occurs, it occurs always after the second input
- Otherwise, you can seemingly use the abbreviation as often as you want without a crash
- The logs show a different Traceback, pointing that issue to be closely related to keyboard restoration.
- The computer never hangs completely, even if switching windows.

Kernel version: 4.19.6.a-1-hardened
Python version: Python 3.7.1

**Comment by luziferius** [comment]
Hm. Can’t reproduce.

I need some more details on how you expand the phrase two times. Do you do it *as fast as possible*?
If that is the case, this might be a multithreading race condition. Two expansion threads run concurrently and try to restore the clipboard. As outlined below, both restore the clipboard and reset the variable. The second thread now sees an empty variable and crashes.

What I tried:
I created a Phrase, set it to paste using the clipboard, added an abbreviation.
I tried t both in autokey-qt and autokey-gtk. I tried with and without the "trigger immediately" option.
In all cases the phrase expanded properly like 10 times in a row (the number of tries I did).

That stack trace you posted is helpful. I know why the lock-up happens, but I have no clue how you got it to crash there like this. The reason is the same as for issue #72. A similar fix to https://github.com/autokey/autokey/pull/162/commits/0afdb9770f614ecb9f848b03b180d537cfe975bc should fix the keyboard lock-up.
```
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 230, in __eventLoop
    method(*args)
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 626, in _restore_clipboard_text
    self.clipboard.text = backup
  File "/usr/lib/python3.7/site-packages/autokey/interface.py", line 161, in text
    self._clipboard.set_text(new_content, -1)
TypeError: Argument 1 does not allow None as a value
```

The “algorithm” explanation as a bit of background :
When you paste using the clipboard, autokey does
- read the current clipboard content and save it into an empty member variable in the Interface class
- build the phrase content and save it to the system clipboard
- Use the keyboard to emit a [Ctrl]+[V] (or other keyboard combinations to cause a paste)
- Copy the previously saved clipboard content back into the system clipboard
- Clear the backup member variable again (to forget the backup, as it won’t be needed any more and thus wastes RAM)

**Comment by LukeLR** [comment]
Oh, thanks for the kind reply! I'm not used to such kindness in open source projects, unfortunately.

No, I'm not using the phrase as fast as possible, the speed does not matter. I have made a quick screencast demonstrating the issue: https://vimeo.com/304689161 On the first run, the phrase works nicely, replacing the abbreviation by the phrase every time. On the second run, it even fails already on the first replacement (haven't had that one yet!), and on the second run, it fails on the second replacement (as described in this issue). In both cases, no keyboard input is received by my window manager anymore. Maybe this helps in understanding the problem a bit better?

**Comment by luziferius** [comment]
Your video indicates it is related to those spurious locking issues… Those were there for ages, like at least 6 years. The original author just silenced them completely.

I tried something. Now, whenever autokey pastes using the clipboard, it will release any grabbed keyboard.
Can you please test, if the patch in this branch https://github.com/autokey/autokey/tree/bugfix/issue_226
fixes the keyboard lock-up issue?
This is the whole diff: https://github.com/autokey/autokey/compare/master...bugfix/issue_226

You can launch the version from the checkout without installing it:
While inside the checkout root dir, do:
```
cd lib
python3 -m autokey.gtkui --verbose # Or put any other CLI switch
```

BTW: you can’t use [Ctrl]+[C] to exit any running autokey instance, because there is no custom signal handler installed. The default action from the GUI toolkit event loop is to just print the stacktrace when it receives a KeyboardInterrupt and continue. Maybe that can be improved…

**Comment by josephj11** [comment]
I'm sure you'll think of something better, but maybe something as simple as meta command (or whatever things like #ifdef are called) would do it. #enable_ctrl_c that would just be used when debugging (programmer's responsibility to remove it - AutoKey wouldn't have to decide when it's valid.) It would be handy if it knew enough to flush the log, etc. before dying.

**Comment by luziferius** [comment]
@LukeLR I pushed another commit into the bugfix branch to change the AutoKey version string. When you test the fix, first check the About dialogue that you started version `0.95.4+bugfix1_issue_225_and_226`

@josephj11 Well, Python internally installs a signal handler for low level POSIX signals like SIGINT, which is sent to a terminal program when [Ctrl]+[C] is pressed. It converts the signal to a KeyboardInterrupt exception and then throws it at the current execution point. The GUI framework event loops just print the stack trace in this case and continues.
We’d need to register a custom SIGINT handler, which calls the shutdown logic when triggered. The best implementation tracks if it was sent twice. On the first event, it calls the shutdown logic, and on the second event, it calls `sys.exit(1)` to terminate the application immediately.
Implementing that handler looks like a nice enhancement.

**Comment by LukeLR** [comment]
Oh, bummer. I have already tested the bugfix branch early this morning, but sent the reply via email and obviously GitHub did not post it here. So I'll copy the contents of my mail this morning here:

> Checked out the branch, and ran your command, still with the same problems. Here's the verbose log of the failure:
>
> 
> ```
> 
> 2018-12-06 08:54:21,196 DEBUG - service - Raw key: ' ', modifiers: [], Key: 
> 2018-12-06 08:54:21,197 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
> 2018-12-06 08:54:21,198 DEBUG - service - No phrase/script matched hotkey
> 2018-12-06 08:54:21,200 DEBUG - iomediator - Send via clipboard
> 2018-12-06 08:54:21,201 DEBUG - interface - Sending string via clipboard: a
> 2018-12-06 08:54:21,201 DEBUG - interface - Sending via clipboard enqueued.
> 2018-12-06 08:54:21,202 DEBUG - service - Input stack at end of handle_keypress: deque([], maxlen=150)
> 2018-12-06 08:54:21,202 ERROR - service - Ignored locking error in handle_keypress
> Traceback (most recent call last):
>   File "/home/lukas/git/02-soft/autokey/lib/autokey/service.py", line 207, in __tryReleaseLock
>     self.configManager.lock.release()
> RuntimeError: release unlocked lock
> 2018-12-06 08:54:21,206 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
> 2018-12-06 08:54:21,209 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
> 2018-12-06 08:54:21,223 DEBUG - iomediator - Send via event interface
> 2018-12-06 08:54:21,225 DEBUG - interface - Send modified key: modifiers: ['<shift>'] key: <insert>
> 2018-12-06 08:54:21,334 ERROR - interface - Error in X event loop thread
> Traceback (most recent call last):
>   File "/home/lukas/git/02-soft/autokey/lib/autokey/interface.py", line 230, in __eventLoop
>     method(*args)
>   File "/home/lukas/git/02-soft/autokey/lib/autokey/interface.py", line 629, in _restore_clipboard_text
>     self.clipboard.text = backup
>   File "/home/lukas/git/02-soft/autokey/lib/autokey/interface.py", line 161, in text
>     self._clipboard.set_text(new_content, -1)
> TypeError: Argument 1 does not allow None as a value
> 2018-12-06 08:54:21,499 DEBUG - service - Raw key: 'b', modifiers: [], Key: b
> 2018-12-06 08:54:21,500 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
> 2018-12-06 08:54:21,501 DEBUG - service - No phrase/script matched hotkey
> 2018-12-06 08:54:21,503 DEBUG - service - Input stack at end of handle_keypress: deque(['b'], maxlen=150)
> 2018-12-06 08:54:21,642 DEBUG - service - Raw key: ' ', modifiers: [], Key: 
> 2018-12-06 08:54:21,642 DEBUG - service - Window visible title: '*Unbenannt 1 - Mousepad', Window class: 'mousepad.Mousepad'
> 2018-12-06 08:54:21,644 DEBUG - service - No phrase/script matched hotkey
> 2018-12-06 08:54:21,647 DEBUG - iomediator - Send via clipboard
> 2018-12-06 08:54:21,647 DEBUG - interface - Sending string via clipboard: a
> 2018-12-06 08:54:21,648 DEBUG - interface - Sending via clipboard enqueued.
> 2018-12-06 08:54:21,649 DEBUG - service - Input stack at end of handle_keypress: deque([], maxlen=150)
> 2018-12-06 08:54:21,650 ERROR - service - Ignored locking error in handle_keypress
> Traceback (most recent call last):
>   File "/home/lukas/git/02-soft/autokey/lib/autokey/service.py", line 207, in __tryReleaseLock
>     self.configManager.lock.release()
> RuntimeError: release unlocked lock
> 2018-12-06 08:54:21,653 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
> 2018-12-06 08:54:21,656 DEBUG - interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
> 
> ```
> 
> Any ideas?

I now issued `git pull` again, and opened autokey with `python3 -m autokey.gtkui --verbose` from the `lib`-folder. ~~In the about-menu, it just says `0.95.4`, although `git status` says:~~
```
:lib$ git status
On branch bugfix/issue_226
Your branch is up to date with 'origin/bugfix/issue_226'.

nothing to commit, working tree clean
```
~~Are you sure that the string in the about menu should have changed? Because the line numbers in the above traceback increased by 3...~~

My fault, autokey was already running and therefore I checked the version in that window. Now, the version string shows: `0.95.4+bugfix1_issue_225_and_226`, but the issue still occurs.

**Comment by luziferius** [comment]
OK, then let's try something else first.
Create an AutoKey script including this:
```
# Release a grabbed keyboard
from Xlib import X, display
local_display = display.Display()
local_display.ungrab_keyboard(X.CurrentTime)
local_display.flush()
```
Open the autokey GUI with this script open, then trigger a keyboard lockup. In the GUI, click the play button in the toolbar to run the script.
(There is a two second delay when using that button.)
Does running the script unlock the keyboard again?

**Comment by LukeLR** [comment]
I created a script with that content, but unfortunately, I can't click the play button after the lockup, because the GUI freezes as well.

But I noticed something else, which might be interesting: The first run of autokey always seems to work as expected, e.g. replaces the abbreviation nearly indefinitely (as in the video). But the second run always freezes at the second replacement (as in the video). Might be coincidence, but I have done multiple testings now and all showed this behaviour...

Might be interesting to figure out if this is part of the bug. I'll check if autokey works better if I wait an hour before the next startup, or if a reboot is required. What content of autokey is restart-persistent? Are there any sockets that don't get closed?

**Comment by luziferius** [comment]
I just got an idea: Does the lock always occur one iteration after that failing `clipboard.set_text()`? It looks like this is a classic deadlock. There is a mutex around that call that isn't released when it fails...

https://github.com/autokey/autokey/blob/b875c2690f45a8776a3bedd3c3ae525bd6e6f5f2/lib/autokey/interface.py#L158-L162

Edit: I pushed another commit to the branch. I tried to mitigate that exceptions during clipboard restoration result in a deadlock. Version string is updated to `bugfix2_issue_225_and_226`.

The persistent content:
- AutoKey does not open any sockets (unless your scripts do ;) ).
- There is the configuration json file located at `$XDG_CONFIG_HOME/autokey/autokey.json`.
- And for each item (Folder, Script, Phrase), there is the file/directory having it’s content and a hidden `.<name_of_item>.json` file containing meta-data, like usage count, assigned hotkeys, abbreviations and all the other settings each item can have.
- There is a log file, containing the INFO and higher level logging output.
- There is a lock file containing the integer PID of the running process, but that one checked for validity, if a stray file is found. If the process noted in the file is not an autokey instance, the file is discarded.
https://github.com/autokey/autokey/blob/b875c2690f45a8776a3bedd3c3ae525bd6e6f5f2/lib/autokey/gtkapp.py#L109-L117

**Comment by luziferius** [comment]
The last commits should fix the issue.

(I’m dumping my thought process here a bit, so this is just a bit of a write-up.)
The None value that causes the clipboard restoration to fail comes from reading the clipboard. There is no other possibility (See below).
I’ve added a warning logger message, if the read clipboard content is None instead of an expected `str` instance. The code now replaces `None` with `""` (empty string). When testing the fix, look out for a warning stating `Tried to backup the X clipboard content, but got None instead of a string.`.

Relevant commit: https://github.com/autokey/autokey/commit/d1d7911060dc1929337f1deda289833311988d3a

For reference:
This block below is the whole clipboard pasting code (sans platform abstraction). The functions are called from top to bottom. In the first, one can ignore the first if branch, dealing with Qt and ignore the inner `if paste_command is None` branch.
First thing `_send_string_clipboard` does is creating a backup of the current clipboard content. That backup is never touched, until it is passed to `_restore_clipboard_text`, which only reads it to write it back to the clipboard. The written property is from the GTK/Qt abstraction layer, which in turn calls the setter quoted in my last comment.

https://github.com/autokey/autokey/blob/d1d7911060dc1929337f1deda289833311988d3a/lib/autokey/interface.py#L599-L641

Thus I conclude the `None` comes from the GTK clipboard getter called in here:

https://github.com/autokey/autokey/blob/d1d7911060dc1929337f1deda289833311988d3a/lib/autokey/interface.py#L151-L156

**Comment by luziferius** [comment]
@LukeLR Ping. Can you try please try the fix again? I think that these measures should do it:

- I guarded the `set_text` calls with a `not None` check to prevent that type error in `_restore_clipboard_text`. This should prevent the application from writing None to the clipboard.
- the `set_text` call itself guards against any exception and always releases the held GTK mutex.
- The `send_string` call that types the paste hotkey guards against BadWindow exceptions and always releases the grabbed keyboard.

**Comment by miskovics** [comment]
Installed the fix, so far no lock up.

**Comment by luziferius** [comment]
@miskovics So you were affected by this bug, too?

**Comment by miskovics** [comment]
I did have frequent lock ups, don't know the root cause. I explained my issue at https://github.com/autokey/autokey/issues/72 .

**Comment by luziferius** [comment]
AutoKey holds volatile references to X11 windows, to be able to match window titles / window classes. Buttons, tooltips, etc are X11 windows too and created/destroyed frequently. If outdated references are accessed, AutoKey receives a BadWindow error.
While AutoKey does keystrokes, it uses the X11 keyboard grabbing mechanism to get sole control over the keyboard, in order to prevent interferences. If a BadWindow happens while AutoKey does stuff with the keyboard, like pressing `[Ctrl]+V`, the normal logic is interrupted. Previously, this caused the keyboard grab release to be skipped, effectively locking you out.

In previous versions I fixed those bugs for the regular typing mode for both the Phrases and Script output. But I didn’t see that it can also happen during pasting using the clipboard. This bugfix branch also includes measures to ensure the release of the keyboard in this case.

**Comment by LukeLR** [comment]
This seems to have fixed the issue for me, although I had to start AutoKey like this in order to run the latest git version: `PYTHONPATH=. python3 -m autokey.gtkui --verbose` within the lib folder, since with `python3 -m autokey.gtkui --verbose` it would still run my local installed version.

**Comment by luziferius** [comment]
Thank you for reporting back. I'll merge the bugfix in the next days.

**Comment by luziferius** [comment]
The fix is in the master branch and will be part of the next release.

**Comment by LukeLR** [comment]
Thanks a lot! I am anticipating this update, and will do further testing ;) Keep up the great work, this is a very cool project!

**Comment by luziferius** [comment]
Released version [`0.95.5`](https://github.com/autokey/autokey/releases/tag/v0.95.5), containing the bug fix.
